language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
             unfoldFixOf (Normalized vPat) (Normalized vArg) k
         _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
 
+
+-- Typecheck a builtin application.  This performs a lot of
+-- substitutions and normalisations, but it should preserve the global
+-- uniqueness condition because it only uses uniquness-preserving
+-- operations from Normalize.Internal.
+
+-- [infer| bn : [a_1 :: K_1, ..., a_n :: K_m](B_1, ..., B_n)C]
+-- [check| G !- A_i :: K_i]
+-- [check| G !- M_j : T_j]     [A_i / a_i]* B_j ~>? S_j   T_j ~ S_j
+-------------------------------------------------------------------
+-- [infer| G !- builtin bn {A_i*} M_j* : NORM ([A_i/a_i]* C)]
+inferTypeM (ApplyBuiltin ann bn typeArgs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  mapping <- checkBuiltinTypeArgs ann bn typeVars typeArgs              -- How to instantiate the type variables
+  normalizedArgTypes <- mapM (substNormalizeMappingM mapping) argTypes --  Instantiate and normalise the argument types
```

Comment:
Do we check anywhere that each type in `typeArgs` (and also `rtype`) are of kind `*`? I guess it's guaranteed by the extensible types machinery, but I think we should perform a check here nonetheless, because the extensible types machinery may change in future and it makes sense to check kinds before normalizing something anyway.

Reply:
`checkBuiltinTypeArgs` does the kind checking for the argument types, but we never check the kind of the return type, not even in the specification.  Should we do that?

I suppose that ideally the (formal) type parameters for a builtin should carry their kinds with them, even if they are all `*`.  I think we're aware that there's a slight mismatch here though.  There was a question of whether reworking the type schemes might help with this.